agent/cert-monitor/config.go

```diff
@@ -0,0 +1,110 @@
+package certmon
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/agent/token"
+	"github.com/hashicorp/consul/tlsutil"
+	"github.com/hashicorp/go-hclog"
+)
+
+// FallbackFunc is used when the normal cache watch based Certificate
+// updating fails to update the Certificate in time and a different
+// method of updating the certificate is required.
+type FallbackFunc func(context.Context) (*structs.SignedResponse, error)
+
+type Config struct {
+	Logger          hclog.Logger
+	TLSConfigurator *tlsutil.Configurator
+	Cache           Cache
+	Tokens          *token.Store
+	Fallback        FallbackFunc
+	FallbackLeeway  time.Duration
+	FallbackRetry   time.Duration
+	InitialCerts    *structs.SignedResponse
+	DNSSANs         []string
+	IPSANs          []net.IP
+	Datacenter      string
+	NodeName        string
+}
+
+// WithCache will cause the created CertMonitor type to use the provided Cache
+func (cfg *Config) WithCache(cache Cache) *Config {
```

Comment:
Any reason you chose to use this "Builder" pattern versus "Functional Options"? I know `auto-config` uses functional options, hence my asking.

Reply:
I kind of go back and forth on this. (up until an hour or two ago this was still using functional options)

My main gripe is with the "options" part of function options. Much of the configuration is not actually optional. It would just be a terrible function signature to required like 7 arguments and then also have functional options for those bits which are actually optional. 

With the config struct I can at least document the required/optional nature of each (it will show up in godoc). The Builder pattern is a remnant of functional options. Instead of scrapping the functions all together I just turned them into config builders.

At the end of all of it, you still have to supply the same number of things, you could inline the config building with the function invocation if you wanted or do it separately. The main difference is really docs which is where functional options fall short.

Reply:
Also I am considering moving AutoConfig over to the same thing.

Reply:
That makes sense to me, I was just thinking a similar thing about the functional options pattern, in that much of it is not actually optional. 👍